FIPS/fip-0032.md

```diff
@@ -21,290 +21,426 @@ replaces: N/A
+syscall gas.
+
+```
+p_extern_gas := 21000
+```
+
+The following syscalls are extern-traversing syscalls, and therefore are subject
+to extern gas:
+
+- `rand::get_chain_randomness`.
+- `rand::get_beacon_randomness`.
+- `crypto::verify_consensus_fault`.
+- `ipld::block_open`.
+
+### IPLD state management gas
+
+Currently, IPLD state block operations incur these fees:
+
+- Reads incur a fixed fee.
+- Writes incur a variable fee, based on the number of bytes written.
+
+With the introduction of the FVM, the state management operations have evolved
+from a two-method design to five syscalls:
+
+- `ipld::block_open`
+- `ipld::block_read`
+- `ipld::block_write`
+- `ipld::block_link`
+- `ipld::block_stat`
+
+For more details, refer to `ipld` namespace syscall catalogue in
+[FIP-0030](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0030.md#namespace-ipld).
+
+**memcpy per byte gas charge**
+
+We introduce a gas charge for memcpy'ing bytes across system boundaries
+(`p_memcpy_gas_per_byte`). The fee is 500 milligas per byte, resulting from
+statistics collected on average memcpy latencies, adapted to the Filecoin gas
+sizing baseline (10gas per ns).
+
+```
+p_memcpy_gas_per_byte := 0.5
+```
+
+**Pricing for `ipld::block_open`**
+
+_Overhead:_ This syscall traverses the FVM <> Node boundary via an extern to
+access the state blockstore. The bytes returned through the extern are copied
+into FVM kernel buffers; thus there is an memcpy and alloc cost (the latter is
+absorbed into the former). The kernel buffers are retained until the actor call
+finisihes.
+
+The pricing formula is:
+
+```
+p_blockopen_base_gas := 114617   # extant fee
+op_ipld_blockopen := p_syscall_gas + p_extern_gas + p_blockopen_base_gas + (p_memcpy_gas_per_byte * block_length)
+```
+
+**Pricing for `ipld::block_read`**
+
+_Overhead:_ This syscall performs an memcpy from the FVM kernel's buffers to
+Wasm memory.
 
-- `get_chain_randomness`: `<<TODO>>`
-- `get_beacon_randomness`: `<<TODO>>`
-- `verify_consensus_fault`: `<<TODO>>`. This is a fixed cost that covers all blockstore operations,
-  signature verifications, and other logic needed to resolve whether a consensus fault has occured.
-  Note that differs from network v15 and earlier, where the cost is variable depending on the exact
-  blockstore operations performed.
-- Blockstore operations: already absorbed in the fees specified above ("IPLD
-  state management fees").
+The pricing formula is:
+
+```
+op_ipld_blockread := p_syscall_gas + (p_memcpy_gas_per_byte * byte_length)
+```
+
+**Pricing for `ipld::block_write`**
+
+_Overhead:_ This syscall performs an memcpy from Wasm memory into the FVM
+kernel's buffers, with a corresponding alloc (absorbed into the memcpy cost).
+The kernel buffers are retained until the actor call finishes.
+
+The pricing formula is:
+
+```
+op_ipld_blockwrite := p_syscall_gas + (p_memcpy_gas_per_byte * byte_length)
+```
+
```

Comment:
Future considerations:

- This call will likely become more expensive when user programmability is introduced as `block_write` will need to reason about data availability.